### DISCUSSION TITLE
API rules for n-m tables.

### DISCUSSION THREAD

**Discussion by newbHy** [pocketbase/pocketbase#1461] [Q&A] [answered]
Hello pocketbase community.

I'm wondering if it's possible to guard api routes when dealing with many-to-many relationships. 

For example i have these three collections:

Rooms:
| id  | name |
| ------------- | ------------- |
| <ID of room 1>  | Some name  |
| <ID of room 2>  | Some other name|

RoomUser:
| room (Relation) | user (Relation) |
| ------------- | ------------- |
| <ID of room 1>  | m7eb1pjwas83xid  |
| <ID of room 1>  | gc5k41x7iqylmr2 |
| <ID of room 2>  | PJKk2iUbESlFI0o |

Message:
| id | user (Relation) | room (Relation)
| ------------- | ------------- | ------------- |
| <ID of msg 1>  | m7eb1pjwas83xid  | <ID of room 1> | 
| <ID of msg 2>  | m7eb1pjwas83xid  | <ID of room 1> | 

Is there any way to apply API rules on the collection Message in such a way that only users from the same room can see each others messages?

For example it shouldn't be possible for PJKk2iUbESlFI0o to retrieve both messages above because he is not listed in the same room as gc5k41x7iqylmr2 or m7eb1pjwas83xid. 

I'm already able to do this through the filter option when using the api endpoints but this wouldn't stop PJKk2iUbESlFI0o from maliciously changing the endpoint url and then he would get access too.

Thanks in advance.

**Comment by ganigeorgiev** [comment, marked as answer]
If I'm understanding correctly your example, I think you should be able to use the `@collection.*` filter in your `message` collection List and View rules:

```js
@collection.roomUser.room = room && @collection.roomUser.user = @request.auth.id
```
The generated SQL query of the above filter would be something like (_simplified version_):

```sql
select *
from messages
left join roomUser
where roomUser.room = messages.room and roomUser.user = {:authId}
```

**Reply by newbHy** [reply]
`@collection.roomUser.room = room && @collection.roomUser.user = @request.auth.id`

This worked perfectly when i applied the id afterwards like this. 

`@collection.roomUser.room.id = room && @collection.roomUser.user.id = @request.auth.id`

@ganigeorgiev Many thanks and especially big thanks for replying so quickly!!

**Reply by ganigeorgiev** [reply]
Since both `roomUser`'s `room` and `user` are single relations, technically both variants, with and without the `.id`, should work (the _`.id` will make one additional join, but I doubt that in your case there will be any noticeable performance difference; this will be auto-optimized internally in the future, so you should use whatever you prefer_).
